bears/python/PEP8NotebookBear.py

```diff
@@ -0,0 +1,131 @@
+        # nbformat.writes does not add a final \n, we do this manually
+        # necessary for the diff because original file has it
+        if self.file[-1].endswith('\n') and not file_lst[-1].endswith('\n'):
+            file_lst[-1] += '\n'
+        return file_lst
+
+    @property
+    def file_nb(self):
+        return copy(self._file_nb)
+
+
+class PEP8NotebookBear(LocalBear):
+    LANGUAGES = {"Python", "Python 2", "Python 3"}
+    REQUIREMENTS = {PipRequirement('autopep8', '1.*'),
+                    PipRequirement('nbformat', '4.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/83333'
+    CAN_FIX = {'Formatting'}
+
+    def run(self, filename, file,
+            max_line_length: int=79,
+            indent_size: int=SpacingHelper.DEFAULT_TAB_WIDTH,
+            pep_ignore: typed_list(str)=(),
+            pep_select: typed_list(str)=(),
+            local_pep8_config: bool=False):
+        """
+        Detects and fixes PEP8 incompliant code in Jupyter Notebooks. This bear
+        will not change functionality of the code in any way.
+
+        :param max_line_length:   Maximum number of characters for a line.
+        :param indent_size:         Number of spaces per indent level.
+        :param pep_ignore:        A list of errors/warnings to ignore.
+        :param pep_select:        A list of errors/warnings to exclusively
+                                  apply.
+        :param local_pep8_config: Set to true if autopep8 should use a config
+                                  file as if run normally from this directory.
+        """
+        options = {"ignore": pep_ignore,
+                   "select": pep_select,
+                   "max_line_length": max_line_length,
+                   "indent_size": indent_size}
+
+        nbhandler = NotebookHandler(filename, file)
+        file_nb = nbhandler.file_nb
+        cells = file_nb['cells']
+
+        @self._bypass_no_newline_at_end_of_file
+        def autopep8_fix_code(source):
+            return autopep8.fix_code(source,
+                                     apply_config=local_pep8_config,
+                                     options=options)
+
+        for i, cell in enumerate(cells):
+            if cell['cell_type'] != 'code':
+                continue
+            source = cell['source']
+            source_corrected = autopep8_fix_code(source)
+            cells[i]['source'] = source_corrected
+
+        corrected = nbhandler.write_to_list(file_nb)
+
+        diffs = Diff.from_string_arrays(file, corrected).split_diff()
+
+        for diff in diffs:
+            yield Result(self,
+                         "The code does not comply to PEP8.",
+                         affected_code=(diff.range(filename),),
+                         diffs={filename: diff})
+
+    @staticmethod
+    def _bypass_no_newline_at_end_of_file(f):
```

Comment:
This is unusual. Why a decorator ? It may make sense to have a contextmanager here ?
Or just put those 2-3 lines in the code itself ... as it is being used only once.